levels/lovelace_tower/events.js

```diff
@@ -11,16 +66,73 @@ const INITIAL_STATE = {
 };
 
 module.exports = async function (event, world) {
-  const worldState = merge(INITIAL_STATE, world.getState(LOVELACE_TOWER_STATE_KEY));
+  const worldState = merge(
+    INITIAL_STATE,
+    world.getState(LOVELACE_TOWER_STATE_KEY)
+  );
+
+  // Taken fom https://github.com/TwilioQuest/twilioquest-dev-fundamentals/blob/main/levels/tower_of_knowledge/events.js#:~:text=%7D-,//%20Match%20objectives%20to%20shelves%20they%20should%20unlock%20when%20an,%7D,-//%20Hide%20entities
+  const unlockPairs = [
+    ["api-01-local-function", "api-hidden-door-1"],
+    ["api-02-async-await", "api-hidden-door-2"],
+    ["api-03-fetch", "api-hidden-door-3"],
+  ];
+
+  const doorPairs = [["api-04-remote-and-local", "api-door-1"]];
+
+  if (
+    event.name === "objectiveCompleted" ||
+    event.name === "objectiveCompletedAgain"
+  ) {
+    unlockPairs.forEach(([objectiveKey, hiddenDoorKey]) => {
+      if (
+        event.objective === objectiveKey &&
+        !worldState.insideLovelaceTower.hiddenEntities.includes(hiddenDoorKey)
+      ) {
+        worldState.insideLovelaceTower.hiddenEntities.push(hiddenDoorKey);
+      }
+    });
+
+    doorPairs.forEach(([objectiveKey, doorKey]) => {
+      if (
+        event.objective === objectiveKey &&
+        !worldState.insideLovelaceTower.openedDoors.includes(doorKey)
+      ) {
+        worldState.insideLovelaceTower.openedDoors.push(doorKey);
+      }
+    });
+  }
+
+  // Hide entities
+  [
+    ...worldState.insideLovelaceTower.hiddenEntities,
+    ...worldState.insideLibrary.hiddenEntities,
+  ].forEach((hiddenEntityKey) => {
+    world.hideEntities(hiddenEntityKey);
+  });
+
+  // Open doors
+  [
+    ...worldState.insideLovelaceTower.openedDoors,
+    ...worldState.insideLibrary.openedDoors,
+  ].forEach((openedDoorKey) => {
+    world.forEachEntities(
+      openedDoorKey,
+      (door) => door.state && door.state.fsm && door.state.fsm.action("open")
+    );
+  });
+
+  if (event.name === "playerDidInteract") {
+    handleSpells(event, world, {
+      ...worldState,
+    });
+  }
 
   console.log(`event: ${event.name}`);
   console.log(`event target ${event.target}`);
   console.log(`event target ${event.target && event.target.key}`);
   console.log(worldState);
 
-// Library exit not accessible until obj5Complete === true
-world.disableTransitionAreas("exit_to_library_corridor");
```

Comment:
Removed this because it was tackled in [another PR](https://github.com/TwilioQuest/twilioquest-api-academy/pull/76) using [property binding](https://twilioquest.github.io/extension-docs/api/custom_objects.html#property-binding)

<img width="1440" alt="property-binding-example" src="https://user-images.githubusercontent.com/51891409/204931281-f0232e10-846c-4c4e-b836-83fa8a0b9a50.png">